src/bin/customer/establish.rs

```diff
@@ -182,179 +116,89 @@ impl Command for Establish {
 
-        // Write out establishment struct to disk if operating in off-chain mode
-        if self.off_chain {
-            write_establish_json(&establishment)?;
-        }
-
-        // The customer and merchant funding information
-        let merchant_funding_info = tezos::MerchantFundingInformation {
-            balance: merchant_balance,
-            address: contract_details.merchant_funding_address(),
-            public_key: contract_details.merchant_tezos_public_key.clone(),
-        };
-        let customer_funding_info = tezos::CustomerFundingInformation {
-            balance: customer_balance,
-            address: tezos_address.clone(),
-            public_key: tezos_public_key.clone(),
-        };
-
-        let (contract_id, origination_status) = if self.off_chain {
-            // TODO: prompt user to submit the origination of the contract
-            todo!("prompt user to submit contract origination details")
-        } else {
-            // Originate the contract on-chain
-            tezos::originate(
-                Some(&config.tezos_uri),
-                &merchant_funding_info,
-                &customer_funding_info,
-                &establishment.merchant_ps_public_key,
-                &tezos_key_material,
-                &channel_id,
-                config.confirmation_depth,
-                config.self_delay,
-            )
-            .await
-            .context("Failed to originate contract on-chain")?
-        };
-
-        // Check to make sure origination succeeded
-        if !matches!(origination_status, tezos::OperationStatus::Applied) {
-            todo!("Abort protocol because origination failed?")
-        }
-
-        // Update database to indicate successful contract origination.
-        database
-            .with_channel_state(
-                &actual_label,
-                zkchannels_state::Inactive,
-                |inactive| -> Result<_, Infallible> { Ok((State::Originated(inactive), ())) },
-            )
-            .await
-            .context(format!(
-                "Failed to update channel {} to Originated status",
-                &actual_label
-            ))??;
-
-        database
-            .initialize_contract_details(&actual_label, &contract_id)
-            .await
-            .context(format!(
-                "Failed to store contract details for {}",
-                &actual_label
-            ))?;
-
-        // Send the contract id to the merchant.
-        let chan = chan
-            .send(contract_id.clone())
-            .await
-            .context("Failed to send contract id to merchant")?;
+        let chan = originate_contract(
+            chan,
+            &channel_name,
+            channel_id,
+            database.as_ref(),
+            &config,
+            &merchant_funding_info,
+            &customer_funding_info,
+            &zkabacus_customer_config,
+            self.off_chain,
+        )
+        .await?;
```

Comment:
I don't feel that separating this code actually makes it more skimmable in practice. At the code level, `fn originate_contract` can do pretty anything that `run()` can do, so I'd want to read them together to have the full picture. 

In fact, I sort of think we should go the other way - I think even the functions that exist solely to add a `with_timeout()` should become inline `async {...}.with_timeout(...)` blocks to keep the protocol logic in one place.

Reply:
Yeah, I can get behind this. The sheer volume of parameters plus the non-reusability of the function makes this a worse solution that I was hoping. I will try adding async blocks instead.

Reply:
This is now done!